### PR TITLE
Add a shared value cache for execution cursors

### PR DESCRIPTION
**Pull request by PlasmaPower**

### CODE REVIEW

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-avm-cpp/data_storage/src/value/valuecache.cpp [hunk 1]
```diff
@@ -77,3 +77,16 @@ void ValueCache::nextCache() {
     saving_cache_index = (saving_cache_index + 1) % caches.size();
     caches[saving_cache_index].clear();
 }
+
+void ValueCache::maybeInitializeFrom(const ValueCache& other) {
+    if (saving_cache_index == 0 && !caches.empty() && caches.back().empty() &&
```
[the remark this answers is not in the document]
Comment (PlasmaPower) [reply packages/arb-avm-cpp/data_storage/src/value/valuecache.cpp:82]: Look at nextCache's implementation -- we're checking if the _previous_ cache is empty, as the current one is easily cleared. Perhaps a better check would be if all caches are empty.

[the remark this answers is not in the document]
Comment (PlasmaPower) [reply packages/arb-avm-cpp/data_storage/src/value/valuecache.cpp:82]: I was only intending this function to run on an empty cache, though perhaps it shouldn't be limited as such.

[the remark this answers is not in the document]
Comment (PlasmaPower) [reply packages/arb-avm-cpp/data_storage/src/value/valuecache.cpp]: I've changed this to not require an empty cache.

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**